File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/exam.ts
/** Exam domain types: exam definitions, lifecycle, variants, bindings. */

import type { BindingSource, BindingStatus, ExamState } from "./enums";

/** Lightweight exam record for list views. */
export interface ExamSummary {
  exam_id: string;
  subject_id: string;
  class_id: string;
  title?: string;
  scheduled_at: string;
  duration_min?: number;
  state: ExamState;
}

/** Full exam configuration including rubric metadata. */
export interface ExamDetail extends ExamSummary {
  section_id: string;
  total_marks: number;
  question_count: number;
  late_entry_cutoff_min?: number;
  objection_window_days?: number;
  variants?: string[];
  created_by: string;
}

/** Request body for creating a new exam. */
export interface CreateExamRequest {
  title: string;
  subject_id: string;
  class_id: string;
  section_id: string;
  scheduled_at: string;
  duration_min: number;
  question_count: number;
  total_marks: number;
  negative_marking?: boolean;
  variants?: string[];
}

/** Partial update for mutable exam fields (before lock). */
export interface PatchExamRequest {
  scheduled_at?: string;
  duration_min?: number;
  objection_window_days?: number;
  late_entry_cutoff_min?: number;
}

/** Request to transition an exam to a new lifecycle state. */
export interface LifecycleTransitionRequest {
  to_state: ExamState;
  actor_id: string;
  reason?: string;
}

/** Result of applying a lifecycle state transition. */
export interface LifecycleTransitionResult {
  exam_id: string;
  from_state: string;
  to_state: string;
  changed_at: string;
}

/** Assign invigilators and evaluators to an exam. */
export interface AssignmentRequest {
  invigilator_ids: string[];
  evaluator_ids: string[];
  double_blind?: boolean;
}

/** Reference to a student from the Stoody roster. */
export interface StudentRef {
  student_id: string;
  name: string;
  roll?: string;
  section_id?: string;
}

/** Request to bind a pen to a student for an exam. */
export interface CreateBindingRequest {
  pen_mac: string;
  student_id: string;
  source: BindingSource;
  request_id?: string;
}

/** Confirm or reject a provisional pen-student binding. */
export interface ConfirmBindingRequest {
  status: BindingStatus;
  rejection_reason?: string;
}

/** Server-side pen-student binding record. */
export interface BindingRecord {
  exam_id: string;
  pen_mac: string;
  student_id?: string;
  student_name?: string;
  student_roll?: string;
  status: BindingStatus;
  source: BindingSource;
  bound_at: string;
  server_confirmed_at?: string;
  rejection_reason?: string;
}

File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/exam.py
"""Exam domain models: exam definitions, lifecycle, variants, rubric, question regions."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import (
    BindingSource,
    BindingStatus,
    ExamState,
)


class ExamSummary(BaseModel):
    """Lightweight exam record for list views."""

    exam_id: UUID
    subject_id: str
    class_id: str
    title: Optional[str] = None
    scheduled_at: datetime
    duration_min: Optional[int] = None
    state: ExamState


class ExamDetail(ExamSummary):
    """Full exam configuration including rubric metadata."""

    section_id: str
    total_marks: float
    question_count: int
    late_entry_cutoff_min: Optional[int] = None
    objection_window_days: Optional[int] = None
    variants: Optional[list[str]] = None
    created_by: str


class CreateExamRequest(BaseModel):
    """Request body for creating a new exam."""

    title: str
    subject_id: str
    class_id: str
    section_id: str
    scheduled_at: datetime
    duration_min: int
    question_count: int
    total_marks: float
    negative_marking: Optional[bool] = None
    variants: Optional[list[str]] = None


class PatchExamRequest(BaseModel):
    """Partial update for mutable exam fields (before lock)."""

    scheduled_at: Optional[datetime] = None
    duration_min: Optional[int] = None
    objection_window_days: Optional[int] = None
    late_entry_cutoff_min: Optional[int] = None


class LifecycleTransitionRequest(BaseModel):
    """Request to transition an exam to a new lifecycle state."""

    to_state: ExamState
    actor_id: str
    reason: Optional[str] = None


class LifecycleTransitionResult(BaseModel):
    """Result of applying a lifecycle state transition."""

    exam_id: UUID
    from_state: str
    to_state: str
    changed_at: datetime


class AssignmentRequest(BaseModel):
    """Assign invigilators and evaluators to an exam."""

    invigilator_ids: list[str]
    evaluator_ids: list[str]
    double_blind: Optional[bool] = None


class StudentRef(BaseModel):
    """Reference to a student from the Stoody roster."""

    student_id: str
    name: str
    roll: Optional[str] = None
    section_id: Optional[str] = None


class CreateBindingRequest(BaseModel):
    """Request to bind a pen to a student for an exam."""

    pen_mac: str
    student_id: str
    source: BindingSource
    request_id: Optional[str] = None


class ConfirmBindingRequest(BaseModel):
    """Confirm or reject a provisional pen-student binding."""

    status: BindingStatus
    rejection_reason: Optional[str] = None


class BindingRecord(BaseModel):
    """Server-side pen-student binding record."""

    exam_id: UUID
    pen_mac: str
    student_id: Optional[str] = None
    student_name: Optional[str] = None
    student_roll: Optional[str] = None
    status: BindingStatus
    source: BindingSource
    bound_at: datetime
    server_confirmed_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None
